service/src/main/java/bio/terra/tanagra/app/AuthInterceptor.java

```diff
@@ -0,0 +1,125 @@
+package bio.terra.tanagra.app;
+
+import bio.terra.common.iam.BearerToken;
+import bio.terra.common.iam.BearerTokenFactory;
+import bio.terra.tanagra.app.configuration.AuthConfiguration;
+import bio.terra.tanagra.exception.SystemException;
+import bio.terra.tanagra.service.auth.BearerTokenUtils;
+import bio.terra.tanagra.service.auth.IapJwtUtils;
+import bio.terra.tanagra.service.auth.InvalidTokenException;
+import bio.terra.tanagra.service.auth.UserAuthentication;
+import bio.terra.tanagra.service.auth.UserId;
+import com.google.api.client.http.HttpMethods;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.Authorization;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+@Service
+public class AuthInterceptor implements HandlerInterceptor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AuthInterceptor.class);
+
+  private final AuthConfiguration authConfiguration;
+
+  @Autowired
+  public AuthInterceptor(AuthConfiguration authConfiguration) {
+    this.authConfiguration = authConfiguration;
+  }
+
+  /**
+   * Returns true if the request is authenticated and can proceed. Publishes authenticated user info
+   * using Spring's SecurityContext.
+   */
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws Exception {
+    // Clear the security context before we start, to make sure we're not using authentication
```

Comment:
I'm not very familiar with Spring so I don't understand the details of how this machinery works, but this is a terrifying comment. Are requests not completely isolated? What happens if there are two or more concurrent requests?

Reply:
I tried to copy the AoU authentication strategy as much as possible. This method in particular is a straight [copy](https://github.com/all-of-us/workbench/blob/main/api/src/main/java/org/pmiops/workbench/interceptors/AuthInterceptor.java#L203). I also spent a while looking into the Spring stuff and I think this is fine. Each request should be handled by a separate thread, and the default mode of this class is thread-local ([ref](https://docs.spring.io/spring-security/site/docs/3.0.x/apidocs/org/springframework/security/core/context/SecurityContextHolder.html)).